### PR TITLE
fix(omnisharp): update on_new_config

### LINKED ISSUE #2392: LSP features not working in second dotnet solution with OmniSharp
### Description

I'm not sure if it's lspconfig-related or omnisharp-related, so posting it here to start with.

### Neovim version

NVIM v0.9.0-dev-774+g151b9fc52-dirty
Build type: Release
LuaJIT 2.1.0-beta3

### Nvim-lspconfig version

85cd2ecacd8805614efe3fb3a5146ac7d0f88a17

### Operating system and version

macOs 13.1

### Affected language servers

omnisharp

### Steps to reproduce

Step 1: open file from solution 1

LspInfo:
```
 Language client log: /Users/starteleport/.local/state/nvim/lsp.log
 Detected filetype:   cs
 
 1 client(s) attached to this buffer: 
 
 Client: omnisharp (id: 1, bufnr: [1])
 	filetypes:       cs, vb
 	autostart:       true
 	root directory:  /Users/starteleport/my_code/solution1
 	cmd:             omnisharp -z -s /Users/starteleport/my_code/solution1 --hostPID 59024 DotNet:enablePackageRestore=false --encoding utf-8 --languageserver FormattingOptions:EnableEditorConfigSupport=true Sdk:IncludePrereleases=true
 
 Configured servers list: omnisharp
```

Step 2: check whether things like goto definition work within solution 1

Step 3: open file from solution 2

LspInfo:
```
 Language client log: /Users/starteleport/.local/state/nvim/lsp.log
 Detected filetype:   cs
 
 1 client(s) attached to this buffer: 
 
 Client: omnisharp (id: 1, bufnr: [1, 7])
 	filetypes:       cs, vb
 	autostart:       true
 	root directory:  /Users/starteleport/my_code/solution2
 	cmd:             omnisharp -z -s /Users/starteleport/my_code/solution1 --hostPID 59024 DotNet:enablePackageRestore=false --encoding utf-8 --languageserver FormattingOptions:EnableEditorConfigSupport=true Sdk:IncludePrereleases=true -z -s /Users/starteleport/my_code/solution2 --hostPID 59024 DotNet:enablePackageRestore=false --encoding utf-8 --languageserver FormattingOptions:EnableEditorConfigSupport=true Sdk:IncludePrereleases=true
 
 Configured servers list: omnisharp
```

Step 4: check whether things like goto definition work within solution 2

Step 5: issue `ps -A` to verify how many instances of `omnisharp` are started and with which parameters

### Actual behavior

Goto definition from step 2 work within solution 1.

Goto definition from step 4 do not work within solution 2.

LspInfo in step 4 show suspicious `cmd` value that expands even further as I open subsequent buffers.

`cmd` value in step 4 does not match the output from `ps -A` in step 5

### Expected behavior

Goto definition from step 2 work within solution 1.

Goto definition from step 4 work within solution 2.

LspInfo in step 4 shows adequate `cmd` value that matches the output of `ps -A`.

### Minimal config

```Lua
local on_windows = vim.loop.os_uname().version:match 'Windows'

local function join_paths(...)
  local path_sep = on_windows and '\\' or '/'
  local result = table.concat({ ... }, path_sep)
  return result
end

vim.cmd [[set runtimepath=$VIMRUNTIME]]

local temp_dir = vim.loop.os_getenv 'TEMP' or '/tmp'

vim.cmd('set packpath=' .. join_paths(temp_dir, 'nvim', 'site'))

local package_root = join_paths(temp_dir, 'nvim', 'site', 'pack')
local lspconfig_path = join_paths(package_root, 'test', 'start', 'nvim-lspconfig')

if vim.fn.isdirectory(lspconfig_path) ~= 1 then
  vim.fn.system { 'git', 'clone', 'https://github.com/neovim/nvim-lspconfig', lspconfig_path }
end

vim.lsp.set_log_level 'trace'
require('vim.lsp.log').set_format_func(vim.inspect)
local nvim_lsp = require 'lspconfig'
local on_attach = function(_, bufnr)
  local function buf_set_option(...)
    vim.api.nvim_buf_set_option(bufnr, ...)
  end

  buf_set_option('omnifunc', 'v:lua.vim.lsp.omnifunc')

  -- Mappings.
  local opts = { buffer = bufnr, noremap = true, silent = true }
  vim.keymap.set('n', 'gD', vim.lsp.buf.declaration, opts)
  vim.keymap.set('n', 'gd', vim.lsp.buf.definition, opts)
  vim.keymap.set('n', 'K', vim.lsp.buf.hover, opts)
  vim.keymap.set('n', 'gi', vim.lsp.buf.implementation, opts)
  vim.keymap.set('n', '<C-k>', vim.lsp.buf.signature_help, opts)
  vim.keymap.set('n', '<space>wa', vim.lsp.buf.add_workspace_folder, opts)
  vim.keymap.set('n', '<space>wr', vim.lsp.buf.remove_workspace_folder, opts)
  vim.keymap.set('n', '<space>wl', function()
    print(vim.inspect(vim.lsp.buf.list_workspace_folders()))
  end, opts)
  vim.keymap.set('n', '<space>D', vim.lsp.buf.type_definition, opts)
  vim.keymap.set('n', '<space>rn', vim.lsp.buf.rename, opts)
  vim.keymap.set('n', 'gr', vim.lsp.buf.references, opts)
  vim.keymap.set('n', '<space>e', vim.diagnostic.open_float, opts)
  vim.keymap.set('n', '[d', vim.diagnostic.goto_prev, opts)
  vim.keymap.set('n', ']d', vim.diagnostic.goto_next, opts)
  vim.keymap.set('n', '<space>q', vim.diagnostic.setloclist, opts)
end

-- Add the server that troubles you here
local name = 'omnisharp'
local cmd = { 'omnisharp' } -- needed for elixirls, omnisharp, sumneko_lua
if not name then
  print 'You have not defined a server name, please edit minimal_init.lua'
end
if not nvim_lsp[name].document_config.default_config.cmd and not cmd then
  print [[You have not defined a server default cmd for a server
    that requires it please edit minimal_init.lua]]
end

nvim_lsp[name].setup {
  cmd = cmd,
  on_attach = on_attach,
}

print [[You can find your log at $HOME/.cache/nvim/lsp.log. Please paste in a github issue under a details tag as described in the issue template.]]
```

### LSP log

https://gist.github.com/starteleport/055fedceaea9e2ef22dd2ff9e6be43d9

### PR DESCRIPTION
**Pull request by JustALawnGnome7**
Fixes #2392

### CODE REVIEW

File: lua/lspconfig/server_configurations/omnisharp.lua [hunk 1]
```diff
@@ -1,4 +1,5 @@
 local util = require 'lspconfig.util'
+local cmd_init = nil
```
Comment (williamboman) [inline lua/lspconfig/server_configurations/omnisharp.lua:2]:
```suggestion

```

File: lua/lspconfig/server_configurations/omnisharp.lua [hunk 2]
```diff
@@ -42,37 +43,44 @@ return {
...
+        for _, v in pairs(new_config.cmd) do
+          table.insert(cmd_init, v)
+        end
+      end
+      -- copy the fields of `cmd_init` into `new_config.cmd`
+      new_config.cmd = {}
+      for _, v in pairs(cmd_init) do
+        table.insert(new_config.cmd, v)
+      end
+      -- add hard-coded command arguments
```
Comment (williamboman) [inline lua/lspconfig/server_configurations/omnisharp.lua:58]: This one-liner effectively does the same thing without having to keep "global" state:

```suggestion
      new_config.cmd = { unpack(new_config.cmd or {}) }
```

File: lua/lspconfig/server_configurations/omnisharp.lua [hunk 3]
```diff
@@ -42,37 +43,44 @@ return {
...
         table.insert(new_config.cmd, 'RoslynExtensionsOptions:EnableImportCompletion=true')
       end
-
       if new_config.sdk_include_prereleases then
         table.insert(new_config.cmd, 'Sdk:IncludePrereleases=true')
       end
-
       if new_config.analyze_open_documents_only then
         table.insert(new_config.cmd, 'RoslynExtensionsOptions:AnalyzeOpenDocumentsOnly=true')
       end
```
Comment (williamboman) [inline lua/lspconfig/server_configurations/omnisharp.lua:86]: Might as well also add this (see 2nd point [here](https://github.com/neovim/nvim-lspconfig/issues/2392#issuecomment-1610129809))?

```suggestion

      new_config.capabilities = vim.deepcopy(new_config.capabilities)
      new_config.capabilities.workspace.workspaceFolders = false -- https://github.com/OmniSharp/omnisharp-roslyn/issues/909
      end
```

File: lua/lspconfig/server_configurations/omnisharp.lua (hunk 2 above)
Comment (JustALawnGnome7) [reply lua/lspconfig/server_configurations/omnisharp.lua]: That's a great suggestion, @williamboman. I made your suggested changes, tested them, and committed them: they work great! Thanks!

File: lua/lspconfig/server_configurations/omnisharp.lua (hunk 3 above)
Comment (JustALawnGnome7) [reply lua/lspconfig/server_configurations/omnisharp.lua:75]: Thanks, I'll take a look at this soon.

File: lua/lspconfig/server_configurations/omnisharp.lua (hunk 2 above)
Comment (JustALawnGnome7) [reply lua/lspconfig/server_configurations/omnisharp.lua]: @williamboman It looks like I may have been bit hasty in my testing. While your suggestion to use `unpack` works when `cmd` is configured as a table with only one item (e.g., `cmd = { "omnisharp" }`), it fails if `cmd` needs to be instantiated in any other way.

For example, I've seen the following code suggestion in numerous places on GitHub, and it's actually included in the `docs` section of the same `omnisharp.lua` file:

``` lua
cmd = { "dotnet", "/path/to/omnisharp/OmniSharp.dll" }
```

In this instance, the line performing the `unpack` will simply assign `{ "dotnet" }` to `new_config.cmd`. Additionally, any other arguments that the end user manually passes into the program will also get dropped.

I don't see how we can accomplish what we're trying to without keeping "global" state as I had done previously, though I'm certainly no Lua expert. I'm definitely up for suggestions! And if I need to double back and stick with my original approach, I'd love to know if there's a more idiomatic way to accomplish what I was originally trying to do.

Comment (williamboman) [reply lua/lspconfig/server_configurations/omnisharp.lua]: Huh, that sounds odd. So the only thing we're trying to do here is copy a list, so no need to change approach entirely.

`new_config.cmd = vim.list_extend({}, new_config.cmd)` is an alternative, and might even be more idiomatic.

Comment (JustALawnGnome7) [reply lua/lspconfig/server_configurations/omnisharp.lua]: My apologies _again_, the issue on my end was a configuration and typographical error. What I currently have committed works properly per your previous suggestion. I appreciate your patience with me!

File: lua/lspconfig/server_configurations/omnisharp.lua (hunk 3 above)
Comment (JustALawnGnome7) [reply lua/lspconfig/server_configurations/omnisharp.lua:75]: Oh, very nice! And it seems that this change also renders the following line&mdash;and function parameter `new_root_dir`&mdash;unnecessary. Would you agree?

``` lua
vim.list_extend(new_config.cmd, { '-s', new_root_dir })
```

I've tested the code with the above line commented out, and it appears to work correctly by starting up a second OmniSharp instance when needed. And it consistently ties new buffers to the appropriate OmniSharp instances.

Comment (williamboman) [reply lua/lspconfig/server_configurations/omnisharp.lua:75]: I guess it's not needed:

```
-s|--source              Solution or directory for OmniSharp to point at (defaults to current directory).
```

**Comment by JustALawnGnome7** [comment]
@williamboman My apologies, I should have marked the PR as a draft following your initial feedback. Anyway, I think my changes are ready to be reviewed.

**Comment by JustALawnGnome7** [comment]
Also, regarding the comment about using a "rebase workflow" in `CONTRIBUTING.md`... is there a practical way for me to consolidate my commits into one before merging, or is that something that can be done in the merge itself? Again, I'm sorry for all the questions: this is the first PR I've ever submitted in which I actually changed code.

**Review by williamboman — APPROVED** [review]
Tested with multiple workspaces, works fine :+1:!

> Also, regarding the comment about using a "rebase workflow" in CONTRIBUTING.md... is there a practical way for me to consolidate my commits into one before merging, or is that something that can be done in the merge itself? Again, I'm sorry for all the questions: this is the first PR I've ever submitted in which I actually changed code.

<details>
<summary>
Long answer which is irrelevant to the PR itself so hidden to not take up space:
</summary>

I believe PRs are configured to be squash merged in this repo, resulting in a single "squashed" commit in the target branch, so the commits in the PR branch doesn't really matter in the end. The squashed commit message usually defaults to the PR title and body. (I personally think the commit linting rules are a bit too strict in this repo for these reasons)

My personal workflow for maintaining feature branches are something along the lines of:

```sh
# bunch of commits...
# 1. Merge latest upstream.
#    Using merge instead of rebase makes resolving conflicts much easier.
$ git fetch upstream
$ git merge upstream/master

# 2. Reset commit history to match upstream.
#    This will keep your working directory (i.e. local changes on the file system)
#    but restore the commit history to equal upstream.
$ git reset upstream/master

# 3. Commit your changes and force push if remote branch already exists.
$ git commit -am "feat: final commit message with all changes"
$ git push -f
```

When the repository is configured to use squash merge, I tend to skip step 2 and 3 entirely, and just push the merge commit that results from step 1. This will produce a non-linear history in the PR branch, but that will all get squashed to a single commit in the target branch when merged anyway.

Hope this helps!

</details>

File: lua/lspconfig/server_configurations/omnisharp.lua [hunk 4]
```diff
@@ -41,41 +41,43 @@ return {
...
       vim.list_extend(new_config.cmd, { '--hostPID', tostring(vim.fn.getpid()) })
       table.insert(new_config.cmd, 'DotNet:enablePackageRestore=false')
       vim.list_extend(new_config.cmd, { '--encoding', 'utf-8' })
       table.insert(new_config.cmd, '--languageserver')
 
+      -- Append configuration-dependent command arguments
       if new_config.enable_editorconfig_support then
         table.insert(new_config.cmd, 'FormattingOptions:EnableEditorConfigSupport=true')
       end
-
```
Comment (glepnir) [inline lua/lspconfig/server_configurations/omnisharp.lua:55]: these blank line can keep ?

**Pull request merged**